cornell-dti/carriage-web
# Refectored Database to account for admin types

## Summary

Basing it off of @Enochen 's auth refactor branch, I have refactored the database to allow the users to specify types of admins: `sds-admin || redrunner-admin`

This pull request also fixes a lot of issues with creating new accounts: driver, admin, and both admin-driver.

## Refactored Database Schema
- The `Admin` model now contains an `isDriver` field that will be set to true if an employee is both a driver and an admin.
- The `Admin` model now also has a `type` field that is a an array which can include the type of admin: `sds-admin` and/or `redrunner-admin`. I think it should be changed to organization instead.

## Revamped Server Routes
- The `Driver Stats` will now send responses wrapped inside a data field for consistency across all the routes.

## Frontend Logic Changes
- If a one selects both `driver` and one or both of `redrunner-admin`/`sds-admin` there will be two requests creating a driver and an admin object with the same id.
- If one want to promote/demote/also make a driver an admin, the frontend has support for that now. Appropriate requests will be send according to this `onSubmit` function:
https://github.com/cornell-dti/carriage-web/blob/12e68f8c74e71689e9ada0fa7bc3c99fa856661b/frontend/src/components/EmployeeModal/EmployeeModal.tsx#L198-L281
- When trying to see a particular employee (admin/driver), instead of using the `useEmployees` function to get all the `drivers` and `admins` and then searching for the said employee, the frontend, will now make a request to the backend for that specific employee (it's easier to do because then we don't have to worry about common types.)
- If an employee is both `admin` and `driver`, their information will be available at the `admins/:id` endpoint on the frontend. It will fetch both the `admin` and the `driver` and combine them.
https://github.com/cornell-dti/carriage-web/blob/12e68f8c74e71689e9ada0fa7bc3c99fa856661b/frontend/src/components/UserDetail/EmployeeDetail.tsx#L201-L234

## Test Plan
- Try adding and editing drivers/admins: new user, promoting a driver to admin, demoting an admin to driver, making a driver admin as well, making an admin driver as well.
- Make sure you can redirect to the said user's profile after all these changes and try refreshing the page.
- Try adding some rides and make sure they are reflected properly in the stats section of the driver.
- Make sure the Analytics page shows correct data.
- Finally, run `npm run test:server` to make sure backend is working properly.
